src/jdk.httpserver/share/classes/sun/net/httpserver/UnmodifiableHeaders.java

```diff
@@ -52,11 +53,12 @@ public boolean containsValue(Object value) {
             return map.get(key);
         }
 
-        public String getFirst (String key) {
-            return map.getFirst(key);
+        public String getFirst(String key) {
+            final var headers = new Headers();
+            map.forEach((k, v) -> headers.add(k, v.get(0)));
+            return headers.getFirst(key);
         }
```

Comment:
Should be possible to just return a reference to the first String in the map rather than instantiating a new Headers and populating it just for the duration of the call.

Reply:
Headers normalizes the key before any get/getFirst, for example Headers line 131:

`public List<String> get(Object key) {
            return map.get(normalize((String)key));
        }`

I assume we want to maintain this behaviour, but there might be a better way than creating this local Headers?

Reply:
Actually, I just noticed that the List<String>'s are still modifiable. The constructor might have to create an intermediate Map<String,List<String>> with unmodifiable Lists and then you create an unmodifiable Map from that. 

As regards the comment above, yes you want to maintain that behavior. Could you normalize the key here and then just return `map.get(key).get(0)` (with suitable null checks)?

Reply:
Argh. Good catch Michael.  So now we'd also have to reimplement valueSet() and get() to wrap things into unmodifiableList... Or could we get away with providing an immutable snapshot instead of an immutable view?

Reply:
Thanks for catching that. The `List<String>` is now wrapped in an unmodifiable map, and some cases were added to the test.